File: scripts/project-harness-runtime.py
#!/usr/bin/env python3
import argparse,hashlib,json,os,shutil,tempfile
from pathlib import Path
def sha(p):return hashlib.sha256(p.read_bytes()).hexdigest()
def tree(root):
 return [{'path':p.relative_to(root).as_posix(),'sha256':sha(p),'size':p.stat().st_size} for p in sorted(x for x in root.rglob('*') if x.is_file())]
def main():
 ap=argparse.ArgumentParser();ap.add_argument('--internal',required=True);ap.add_argument('--external',required=True);ap.add_argument('--runtime-root',required=True);a=ap.parse_args();internal=Path(a.internal).resolve();external=Path(a.external).resolve();runtime=Path(a.runtime_root).resolve();runtime.mkdir(parents=True,exist_ok=True);partial=Path(tempfile.mkdtemp(prefix='runtime-partial-',dir=runtime));root=partial/'root';root.mkdir()
 try:
  for name in ('.agents','.goose'):
   source=internal/name
   if source.exists():shutil.copytree(source,root/name)
  skills=root/'.agents/skills';skills.mkdir(parents=True,exist_ok=True)
  for item in external.iterdir():
   if item.name=='resolved.json':continue
   shutil.copytree(item,skills/item.name)
  files=tree(root);payload=json.dumps(files,sort_keys=True,separators=(',',':')).encode();digest=hashlib.sha256(payload).hexdigest();manifest={'schema':'harness-runtime-projection-v1','digest':digest,'target':'linux-x86_64','internalManifestSha256':sha(internal/'build-manifest.json'),'externalManifestSha256':sha(external/'resolved.json'),'files':files};(root/'.harness-runtime.json').write_text(json.dumps(manifest,sort_keys=True,indent=2)+'\n');target=runtime/'releases'/digest;target.parent.mkdir(parents=True,exist_ok=True)
  if target.exists():shutil.rmtree(partial)
  else:root.rename(target);shutil.rmtree(partial,ignore_errors=True)
  (runtime/'candidate').write_text(digest+'\n');print(digest)
 except Exception:shutil.rmtree(partial,ignore_errors=True);raise
if __name__=='__main__':main()

File: src/app/harness-manager/src/project-harness-runtime.ts
#!/usr/bin/env node
import { createHash } from "node:crypto";
import { cpSync, existsSync, mkdirSync, mkdtempSync, readdirSync, readFileSync, renameSync, rmSync, statSync, writeFileSync } from "node:fs";
import { basename, join, relative, resolve, sep } from "node:path";

export interface RuntimeFile { path: string; sha256: string; size: number }
export interface ProjectionManifest {
  schema: "harness-runtime-projection-v1";
  digest: string;
  target: "linux-x86_64";
  internalManifestSha256: string;
  externalManifestSha256: string;
  files: RuntimeFile[];
}

export function sha256File(path: string): string {
  return createHash("sha256").update(readFileSync(path)).digest("hex");
}

function comparePaths(a: string, b: string): number { return a < b ? -1 : a > b ? 1 : 0; }

export function runtimeTree(root: string): RuntimeFile[] {
  const paths: string[] = [];
  const walk = (directory: string): void => {
    for (const entry of readdirSync(directory, { withFileTypes: true })) {
      const path = join(directory, entry.name);
      if (entry.isDirectory()) walk(path);
      else if (entry.isFile()) paths.push(path);
    }
  };
  walk(root);
  return paths.sort(comparePaths).map((path) => ({
    path: relative(root, path).split(sep).join("/"),
    sha256: sha256File(path),
    size: statSync(path).size,
  }));
}

function stableJson(value: unknown, indent?: number): string {
  const ordered = (item: unknown): unknown => {
    if (Array.isArray(item)) return item.map(ordered);
    if (item !== null && typeof item === "object") {
      return Object.fromEntries(Object.entries(item).sort(([a], [b]) => comparePaths(a, b)).map(([key, child]) => [key, ordered(child)]));
    }
    return item;
  };
  return JSON.stringify(ordered(value), null, indent);
}

export function projectHarnessRuntime(internalInput: string, externalInput: string, runtimeInput: string): string {
  const internal = resolve(internalInput);
  const external = resolve(externalInput);
  const runtime = resolve(runtimeInput);
  mkdirSync(runtime, { recursive: true });
  const partial = mkdtempSync(join(runtime, "runtime-partial-"));
  const root = join(partial, "root");
  mkdirSync(root);
  try {
    for (const name of [".agents", ".goose"]) {
      const source = join(internal, name);
      if (existsSync(source)) cpSync(source, join(root, name), { recursive: true });
    }
    const skills = join(root, ".agents", "skills");
    mkdirSync(skills, { recursive: true });
    for (const item of readdirSync(external, { withFileTypes: true })) {
      if (item.name === "resolved.json") continue;
      if (!item.isDirectory()) throw new Error("external runtime entry is not a directory: " + item.name);
      cpSync(join(external, item.name), join(skills, item.name), { recursive: true });
    }
    const files = runtimeTree(root);
    const payload = stableJson(files);
    const digest = createHash("sha256").update(payload).digest("hex");
    const manifest: ProjectionManifest = {
      schema: "harness-runtime-projection-v1",
      digest,
      target: "linux-x86_64",
      internalManifestSha256: sha256File(join(internal, "build-manifest.json")),
      externalManifestSha256: sha256File(join(external, "resolved.json")),
      files,
    };
    writeFileSync(join(root, ".harness-runtime.json"), stableJson(manifest, 2) + "\n");
    const releases = join(runtime, "releases");
    const target = join(releases, digest);
    mkdirSync(releases, { recursive: true });
    if (existsSync(target)) rmSync(partial, { recursive: true });
    else {
      renameSync(root, target);
      rmSync(partial, { recursive: true, force: true });
    }
    writeFileSync(join(runtime, "candidate"), digest + "\n");
    return digest;
  } catch (error) {
    rmSync(partial, { recursive: true, force: true });
    throw error;
  }
}

function option(args: string[], name: string): string {
  const index = args.indexOf(name);
  const value = args[index + 1];
  if (index < 0 || value === undefined || value.startsWith("--")) throw new Error("the following arguments are required: " + name);
  return value;
}

export function projectCli(args: string[]): number {
  try {
    const digest = projectHarnessRuntime(option(args, "--internal"), option(args, "--external"), option(args, "--runtime-root"));
    console.log(digest);
    return 0;
  } catch (error) {
    console.error(error instanceof Error ? error.message : String(error));
    return 1;
  }
}

if (process.argv[1] !== undefined && basename(process.argv[1]).replace(/\.(?:js|ts)$/, "") === "project-harness-runtime") process.exitCode = projectCli(process.argv.slice(2));
